packages/pkl.tar/File.pkl

```diff
@@ -0,0 +1,407 @@
+//===----------------------------------------------------------------------===//
+// Copyright © 2025 Apple Inc. and the Pkl project authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+abstract module pkl.tar.File
+
+import "pkl:math"
+
+import "File.pkl"
+import "tar.pkl"
+
+local typealias Buffer = List<UInt8>
+local typealias PosixPortableString =
+  String(
+    codePoints.every((char) ->
+      char == 0 || char.isBetween(0x0007, 0x000d) || char.isBetween(0x0020, 0x007e)
+    )
+  )
+local const blockLength = 512
+local const magic = List(0x75, 0x73, 0x74, 0x61, 0x72, 0x00) // "ustar\0"
+local const version = List(0x30, 0x30) // "00"
+local typealias Block = Buffer(length == blockLength)
+
+/// Pax archive extended header data fields.
+typealias PaxFields = Mapping<PosixPortableString(!contains("=")), String>
+
+local const ustarMaxLinkLength: UInt = 100
+local const ustarMaxUidGid: UInt = 2 ** 21 as UInt
+local const ustarMaxSizeTimestamp: UInt = 2 ** 33 as UInt
+
+/// File type marker, also known as `linkflag`.
+fixed type: Char
+
+/// The numeric byte value of [type].
+fixed typeByte: UInt8 = type.encodeToBytes("UTF-8")[0]
+
+/// File information.
+info: Info?
+
+/// Pax archive extended header data fields.
+paxFields: PaxFields = new {
+  when (info != null) {
+    when (info.gid > ustarMaxUidGid) {
+      ["gid"] = info.gid.toString()
+    }
+    when (info.uid > ustarMaxUidGid) {
+      ["uid"] = info.uid.toString()
+    }
+    when (info.mtime != info.mtime.truncate()) {
+      ["mtime"] = formatPaxNumber(info.mtime)
+    }
+    when (!(info.uname is PosixPortableString)) {
+      ["uname"] = info.uname
+    }
+    when (!(info.gname is PosixPortableString)) {
+      ["gname"] = info.gname
+    }
+  }
+}
+
+/// The payload bytes of this file, padded to the nearest 512 bytes.
+fixed paddedBytes: Buffer(length % blockLength == 0) = List()
```

Comment:
```suggestion
abstract fixed paddedBytes: Buffer(length % blockLength == 0)
```

Reply:
Hmm. For several of the subclasses (`Link`, `SymbolicLink`, and `Directory` specifically), this would just return an empty list, which would mean this gets repeated a bunch.